lib/upperkut/strategies/buffered_queue.rb

```diff
@@ -7,13 +7,65 @@ module Strategies
     class BufferedQueue < Upperkut::Strategies::Base
       include Upperkut::Util
 
+      DEQUEUE_ITEMS = %(
+        local key = KEYS[1]
+        local waiting_ack_key = KEYS[2]
+        local batch_size = ARGV[1]
+        local current_timestamp = ARGV[2]
+        local expired_ack_timestamp = ARGV[3] + 1
+
+        -- move expired items back to the queue
```

Comment:
I liked the idea, but I think we're going to hurt performance doing it for every dequeue op.

Sidekiq does that in the process booting of any worker. Is this not a better idea?
Once the problem is when a Upperkut process is killed, the solution could rely on the start of a fresh new one.

I think the way is implemented here we're unnecessarily adding an extra burden for dequeue ops not affected by the issue we're trying to solve.

Reply:
Good question.

There are some conceptual differences that we need to take into account. The `sidekiq` processing queue and the `upperkut` message acknowledgment are meant for different things. Sidekiq's processing queue ensures that items from a dead worker always get back to the queue, but `upperkut` message acknowledgment ensures that items that, for any reason, do not get `acked` on time will get fetched by another worker as soon as its ack time limit expires. There is not a requirement for a worker to die in order to a message get back to the queue.

There are some improvements made [https://github.com/ResultadosDigitais/upperkut/pull/62#discussion_r409104902] that I think will compensate for the extra burden. What do you think about giving it and performing some load tests to ensure its performance is ok?

Reply:
Make sense.

I think you have a great context (clustering engine) that we can observe performance in a high throughput application. I think this proposal is so much simpler than Sidekiq's one, and we can stick with it, for now. My main concern is just not to slamming Redis with bunch operations unnecessarily.